### ISSUE TITLE
I cannot achieve the result reported in paper with Res2NeXt-29

### ISSUE THREAD

**Issue by c747903646** [gasvn/Res2Net#14]
I tried to implement Res2NeXt-29, 8cx25wx4s by pytorch but could only got classification accuracy 82.32% instead of 83.07% reported in this paper on the CIFAR-100. This error was resulted by randomly initializing parameters of network? 

And the detail of architecture is as below. It's correct?

CifarRes2NeXt(
  (conv_1_3x3): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
  (bn_1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (stage_1): Sequential(
    (stage_1_bottleneck_0): ResNeXtBottleneck(
      (conv_reduce): Conv2d(64, 800, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (convs): ModuleList(
        (0): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(800, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential()
    )
    (stage_1_bottleneck_1): ResNeXtBottleneck(
      (conv_reduce): Conv2d(64, 800, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (convs): ModuleList(
        (0): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(800, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential()
    )
    (stage_1_bottleneck_2): ResNeXtBottleneck(
      (conv_reduce): Conv2d(64, 800, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (convs): ModuleList(
        (0): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(200, 200, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(800, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential()
    )
  )
  (stage_2): Sequential(
    (stage_2_bottleneck_0): ResNeXtBottleneck(
      (conv_reduce): Conv2d(64, 1600, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(1600, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (pool): AvgPool2d(kernel_size=3, stride=2, padding=1)
      (convs): ModuleList(
        (0): Conv2d(400, 400, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(400, 400, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(400, 400, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(1600, 128, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential(
        (shortcut_conv): Conv2d(64, 128, kernel_size=(1, 1), stride=(2, 2), bias=False)
        (shortcut_bn): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
    (stage_2_bottleneck_1): ResNeXtBottleneck(
      (conv_reduce): Conv2d(128, 1600, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(1600, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (convs): ModuleList(
        (0): Conv2d(400, 400, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(400, 400, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(400, 400, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(1600, 128, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential()
    )
    (stage_2_bottleneck_2): ResNeXtBottleneck(
      (conv_reduce): Conv2d(128, 1600, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(1600, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (convs): ModuleList(
        (0): Conv2d(400, 400, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(400, 400, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(400, 400, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(400, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(1600, 128, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential()
    )
  )
  (stage_3): Sequential(
    (stage_3_bottleneck_0): ResNeXtBottleneck(
      (conv_reduce): Conv2d(128, 3200, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(3200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (pool): AvgPool2d(kernel_size=3, stride=2, padding=1)
      (convs): ModuleList(
        (0): Conv2d(800, 800, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(800, 800, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(800, 800, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(3200, 256, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential(
        (shortcut_conv): Conv2d(128, 256, kernel_size=(1, 1), stride=(2, 2), bias=False)
        (shortcut_bn): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
    (stage_3_bottleneck_1): ResNeXtBottleneck(
      (conv_reduce): Conv2d(256, 3200, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(3200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (convs): ModuleList(
        (0): Conv2d(800, 800, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(800, 800, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(800, 800, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(3200, 256, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential()
    )
    (stage_3_bottleneck_2): ResNeXtBottleneck(
      (conv_reduce): Conv2d(256, 3200, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_reduce): BatchNorm2d(3200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (convs): ModuleList(
        (0): Conv2d(800, 800, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (1): Conv2d(800, 800, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
        (2): Conv2d(800, 800, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=8, bias=False)
      )
      (bns): ModuleList(
        (0): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (2): BatchNorm2d(800, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv_expand): Conv2d(3200, 256, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn_expand): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (shortcut): Sequential()
    )
  )
  (classifier): Linear(in_features=256, out_features=100, bias=True)
)

**Comment by gasvn** [comment]
Here is one of my old implementations of res2next on cifar. I have searched my files and this version is the newest version that I can find. Sorry about that. It may have some differences in details with the released version on ImageNet.  According to the training log, it should achieve ~83% top1 acc. on cifar100 with 300 epochs training.
Tips:
Check the parameters of your baseline. And keep the parameters of res2next constant with the baseline to see if there is a performance gain.

https://gist.github.com/gasvn/cd7653ef93fb147be05f1ae4abad6589

**Comment by c747903646** [comment]
I have respectively trained Res2Next-29_6cx24wx4s and Res2Next-29_8cx25wx4s based on this code with SGD optimizer(momentum=0.9, weight-decay=5e-4) and initial rate is 0.1( * 0.1 while epoch = 150 or 225) for 300 epochs training. Training data and testing data were normalized and augmented with codes as below:

`transform_train = transforms.Compose([
            transforms.RandomResizedCrop(224),
            transforms.RandomHorizontalFlip(),
            transforms.ToTensor(),
            transforms.Normalize((0.485, 0.456, 0.406),
                                 (0.229, 0.224, 0.225)),
        ])
        transform_test = transforms.Compose([
            transforms.Resize(args.batch_size),
            transforms.CenterCrop(224),
            transforms.ToTensor(),
            transforms.Normalize((0.485, 0.456, 0.406),
                                 (0.229, 0.224, 0.225)),
        ])`

As a result, Res2Next-29_6cx24wx4s has achieved 82.17 top1 acc on cifar100  and Res2Next-29_8cx25wx4s was 82.29. Which parameter should be change or what process method should be used  could I get the result in paper?

**Comment by gasvn** [comment]
Can you reproduce the performance of ResNeXt? 
I only modified the model file of ResNeXt and do no tuning on hyperparameters.
My conf:
lr = 0.05
wd = 5e-4
mean=[0.485, 0.456, 0.406],
std=[0.229, 0.224, 0.225]

**Comment by qiangwang57** [comment]
> I have respectively trained Res2Next-29_6cx24wx4s and Res2Next-29_8cx25wx4s based on this code with SGD optimizer(momentum=0.9, weight-decay=5e-4) and initial rate is 0.1( * 0.1 while epoch = 150 or 225) for 300 epochs training. Training data and testing data were normalized and augmented with codes as below:
> 
> `transform_train = transforms.Compose([`
> `transforms.RandomCrop(32, padding=4),`
> `transforms.RandomHorizontalFlip(),`
> ` transforms.ToTensor(),`
> ` transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),`
> `])`
> `transform_test = transforms.Compose([`
> ` transforms.ToTensor(),`
> ` transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),`
> ` ])`
> 
> As a result, Res2Next-29_6cx24wx4s has achieved 82.17 top1 acc on cifar100 and Res2Next-29_8cx25wx4s was 82.29. Which parameter should be change or what process method should be used could I get the result in paper?

Hi @c747903646 ,

Have you managed to reproduce the results on cifar100? If so, could you share the code?

Best wishes

Qiang
